include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -2934,6 +2934,347 @@ namespace Differentiation
+     * in both serial and MPI routines.
+     *
+     * @author Jean-Paul Pelteret, 2016, 2017, 2018
+     */
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperScalarFunction
+      : public ADHelperPointLevelFunctionsBase<dim,
+                                               ADNumberTypeCode,
+                                               ScalarType>
+    {
+    public:
+      /**
+       * Type definition for the floating point number type that is used in,
+       * and results from, all computations.
+       */
+      using scalar_type =
+        typename ADHelperBase<ADNumberTypeCode, ScalarType>::scalar_type;
+
+      /**
+       * Type definition for the auto-differentiation number type that is used
+       * in all computations.
+       */
+      using ad_type =
+        typename ADHelperBase<ADNumberTypeCode, ScalarType>::ad_type;
+
+      /**
+       * @name Constructor / destructor
+       */
+      //@{
+
+      /**
+       * The constructor for the class.
+       *
+       * @param[in] n_independent_variables The number of independent variables
+       * that will be used in the definition of the functions that it is
+       * desired to compute the sensitivities of. In the computation of
+       * $\mathbf{f}(\mathbf{X})$, this will be the number of inputs
+       * $\mathbf{X}$, i.e. the dimension of the domain space.
+       */
+      ADHelperScalarFunction(const unsigned int n_independent_variables);
+
+      /**
+       * Destructor
+       */
+      virtual ~ADHelperScalarFunction() = default;
+
+      //@}
+
+      /**
+       * @name Dependent variables
+       */
+      //@{
+
+      /**
+       * Register the definition of the scalar field $\Psi(\mathbf{X})$.
+       *
+       * @param[in] func The recorded function that defines a dependent
+       * variable.
+       *
+       * @note For this class that expects only one dependent variable, this
+       * function must only be called once per tape.
+       *
+       * @note For taped AD numbers, this operation is only valid in recording mode.
+       */
+      void
+      register_dependent_variable(const ad_type &func);
+
+      /**
+       * Compute the value of the scalar field $\Psi(\mathbf{X})$ using the
+       * tape as opposed to executing the source code.
+       *
+       * @return A scalar object with the value for the scalar field evaluated
+       * at the point defined by the independent variable values.
+       */
+      scalar_type
+      compute_value() const;
+
+      /**
+       * Compute the gradient (first derivative) of the scalar field with
+       * respect to all independent variables, i.e.
+       * @f[
+       *   \frac{\partial\Psi(\mathbf{X})}{\partial\mathbf{X}}
+       * @f]
+       *
+       * @return A Vector with the values for the scalar field gradient evaluated
```

Comment:
```suggestion
       * @param[out] gradient A Vector with the values for the scalar field gradient evaluated
```

Is there a precondition that the vector already has the right size?

Reply:
No, we set its size correctly if required.